include/openmc/mesh.h

```diff
@@ -47,39 +44,106 @@ class RegularMesh {
   //! \param[in] p Particle to check
   //! \param[out] bins Bins that were crossed
   //! \param[out] lengths Fraction of tracklength in each bin
-  void bins_crossed(const Particle* p, std::vector<int>& bins,
-                    std::vector<double>& lengths) const;
+  virtual void bins_crossed(const Particle* p, std::vector<int>& bins,
+                            std::vector<double>& lengths) const = 0;
 
   //! Determine which surface bins were crossed by a particle
   //
   //! \param[in] p Particle to check
   //! \param[out] bins Surface bins that were crossed
-  void surface_bins_crossed(const Particle* p, std::vector<int>& bins) const;
+  virtual void
+  surface_bins_crossed(const Particle* p, std::vector<int>& bins) const = 0;
 
   //! Get bin at a given position in space
   //
   //! \param[in] r Position to get bin for
   //! \return Mesh bin
-  int get_bin(Position r) const;
+  virtual int get_bin(Position r) const = 0;
 
   //! Get bin given mesh indices
   //
   //! \param[in] Array of mesh indices
   //! \return Mesh bin
-  int get_bin_from_indices(const int* ijk) const;
+  virtual int get_bin_from_indices(const int* ijk) const = 0;
 
   //! Get mesh indices given a position
   //
   //! \param[in] r Position to get indices for
   //! \param[out] ijk Array of mesh indices
   //! \param[out] in_mesh Whether position is in mesh
-  void get_indices(Position r, int* ijk, bool* in_mesh) const;
+  virtual void get_indices(Position r, int* ijk, bool* in_mesh) const = 0;
 
   //! Get mesh indices corresponding to a mesh bin
   //
   //! \param[in] bin Mesh bin
   //! \param[out] ijk Mesh indices
-  void get_indices_from_bin(int bin, int* ijk) const;
+  virtual void get_indices_from_bin(int bin, int* ijk) const = 0;
+
+  //! Get the number of mesh cells.
+  virtual int n_bins() const = 0;
+
+  //! Get the number of mesh cell surfaces.
+  virtual int n_surface_bins() const = 0;
+
+  //! Find the mesh lines that intersect an axis-aligned slice plot
+  //
+  //! \param[in] plot_ll The lower-left coordinates of the slice plot.
+  //! \param[in] plot_ur The upper-right coordinates of the slice plot.
+  //! \return A pair of vectors indicating where the mesh lines lie along each
+  //!   of the plot's axes.  For example an xy-slice plot will get back a vector
+  //!   of x-coordinates and another of y-coordinates.  These vectors may be
+  //!   empty for low-dimensional meshes.
+  virtual std::pair<std::vector<double>, std::vector<double>>
+  plot(Position plot_ll, Position plot_ur) const = 0;
+
+  //! Write mesh data to an HDF5 group
+  //
+  //! \param[in] group HDF5 group
+  virtual void to_hdf5(hid_t group) const = 0;
+
+  // Data members
+
+  int id_ {-1};  //!< User-specified ID
+  int n_dimension_; //!< Number of dimensions
```

Comment:
I'm hesitant to make `shape_` common because it might not make sense for future `Mesh` subclasses.  `shape_` implies a regular array where `shape_[0] * shape_[1] * shape[2]` equals the number of mesh cells.  But that wouldn't work for a `HexMesh`, for example.

Reply:
Ah, yeah that makes sense. Good thought.